pkg/datastore/headers.go

```diff
@@ -0,0 +1,433 @@
+	}
+
+	if s.hc.UnlinkElement(slot, rev) {
+		return nil // success
+	}
+
+	// revert the saveHeaders operation as revision changed
+	_ = s.Badger.Update(func(txn *badger.Txn) error {
+		return txn.Delete(HeaderKey(slot).Bytes())
+	})
+
+	return errors.New("revision changed")
+}
+
+func putHeaders(ctx context.Context, s *Datastore, slot uint64, cont []structs.HeaderAndTrace, ttl time.Duration) error {
+	buff := bytes.NewBuffer(nil)
+	buff.WriteString("[")
+
+	enc := json.NewEncoder(buff)
+
+	for i, c := range cont {
+		if i > 0 {
+			buff.WriteString(",")
+		}
+		if err := enc.Encode(c); err != nil {
+			return err
+		}
+	}
+	buff.WriteString("]")
+	if err := s.TTLStorage.PutWithTTL(ctx, HeaderKey(slot), buff.Bytes(), ttl); err != nil {
+		return err
+	}
+
+	buff.Truncate(0) // immediately remove
+	return nil
+}
+
+type LoadItem struct {
+	Time    uint64
+	Content []byte
+}
+
+// FixOrphanHeaders is reading all the orphan headers from
+func (s *Datastore) FixOrphanHeaders(ctx context.Context, ttl time.Duration) error {
+	exists := make(map[uint64][]LoadItem)
+
+	// Get all headers, rebuild
+	err := s.Badger.View(func(txn *badger.Txn) error {
+		opts := badger.DefaultIteratorOptions
+		opts.PrefetchValues = false
+
+		it := txn.NewIterator(badger.DefaultIteratorOptions)
+		defer it.Close()
+
+		prefix := []byte("/" + HeaderContentPrefix)
+		re := regexp.MustCompile(`\/hc\/([^\/]+)\/([^\/]+)`)
+
+		for it.Seek(prefix); it.ValidForPrefix(prefix); it.Next() {
+			item := it.Item()
+			subM := re.FindSubmatch(item.Key())
+			if len(subM) != 3 {
+				continue
+			}
+			slot, err := strconv.ParseUint(string(subM[2]), 10, 64)
+			if err != nil {
+				continue
+			}
+			is, ok := exists[slot]
+			if !ok {
+				_, err := txn.Get(append(HeaderPrefixBytes, subM[1]...))
+				if err != nil {
+					if !errors.Is(err, badger.ErrEmptyKey) {
+						return err
+					}
+					exists[slot] = nil
+				}
+				is = []LoadItem{}
+				exists[slot] = is
+			}
+
+			if ok && is == nil {
+				continue
+			}
+
+			li := LoadItem{
+				Time:    item.Version(),
+				Content: make([]byte, item.ValueSize())}
+			item.ValueCopy(li.Content)
+			exists[slot] = append(is, li)
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	buff := new(bytes.Buffer)
+	for slot, v := range exists {
+		if v != nil {
+			buff.Reset()
+			sort.Slice(v, func(i, j int) bool {
+				return v[i].Time > v[j].Time
+			})
+
+			buff.WriteString("[")
+			for i, payload := range v {
+				if i > 0 {
+					buff.WriteString(",")
+				}
+				io.Copy(buff, bytes.NewReader(payload.Content))
+
+			}
+			buff.WriteString("]")
+
+			if err := s.TTLStorage.PutWithTTL(ctx, HeaderKey(slot), buff.Bytes(), ttl); err != nil {
+				return err
+			}
+		}
+	}
+	return err
+}
+
+func (s *Datastore) MemoryCleanup(ctx context.Context, ttl time.Duration) error {
+	for {
+		time.Sleep(InMemorySlotPurgeCheck)
+		slots, ok := s.hc.CheckForRemoval()
+		if !ok {
+			continue
+		}
+
+		s.SaveHeaders(ctx, slots, ttl)
```

Comment:
After storing headers, where are the memory items being removed?

Reply:
https://github.com/blocknative/dreamboat/blob/0b5d6988e24d15955eb28ab95ed24b1e74d70c5e/pkg/datastore/headers.go#L302